docs/input_output.md

```diff
@@ -2,85 +2,279 @@
+          useQueue: true
+          stack: '$.cumulus_meta.stack'
+          provider: '$.meta.provider'
+          inlinestr: 'prefix{meta.foo}suffix',
+          array: '{[$.meta.foo]}',
+          object: '{{$.meta}}'
+
+The task configuration in the message (only partial message is shown here):
+
+    {
+      "cumulus_meta": {
+        "stack": "foo-cumulus",
+        ....
+      },
+      "meta": {
+        "foo": "bar",
+        "provider": {
+          "id": "FOO_DAAC",
+          "anykey": "anyvalue"
+        },
+        ...
+      },
+      "workflow_config": {
+        "Discovery": {
+          "useQueue": true,
+          "stack": "{{$.cumulus_meta.stack}}",
+          "object": "{{$.meta.provider}}",
+          "inlinestr": "prefix{meta.foo}suffix",
+          "array": "{[$.meta.foo]}"
+        },
+        ...
       }
     }
 
-When a Task receives such a payload, it is responsible for fetching the JSON document at the given key and replacing `"payload"` with its contents.  See the (EventSource)[../lib/event-source.js] implementation for examples.
+Into this as part of the message passed to task:
+
+    {
+      "config" : {
+        "useQueue": true,
+        "stack": "foo-cumulus",
+        "object": {
+          "id": "FOO_DAAC",
+          "anykey": "anyvalue"
+        },
+        "inlinestr": "prefixbarsuffix",
+        "array": ["bar"]
+      }
+      
+URL template variables replace dotted paths inside curly brackets with their corresponding value. If the Cumulus Message Adapter cannot resolve a value, it should ignore the template, leaving it verbatim in the string.  While seemingly complex, this allows significant decoupling of Tasks from one another and the data that drives them. Tasks are able to easily receive runtime configuration produced by previously run Tasks and domain data.
 
-## URL Templating
+### 3. Resolve task input
 
-When each task executes, it is expected to resolve URL templates found in its collection configuration against the entire collection configuration. For example, tasks should resolve the following collection:
+By default, the incoming payload is the payload from previous task.  The task can also be configured to use a portion of the payload its input message.  For example, a task specifies cumulus_message.input in its workflow configuration:
+    
+    ExampleTask:
+      config:
+        cumulus_message:
+            input: '$.payload.foo'
+            
+The task configuration in the message:
 
     {
-      "meta": { "name": "Hello" },
-      "config" : { "output" : "{meta.name} World!" }
+      "workflow_config": {
+        "ExampleTask": {
+          "cumulus_message": {
+            "input": "{{$.payload.foo}}"
+          }
+        }
+      },
+      "payload": {
+        "foo": {
+          "anykey": "anyvalue"
+        }
+      }
     }
 
-Into this:
+The Cumulus Message Adapter will resolve the task input, instead of send the whole `"payload"` as task input, the task input would be:
```

Comment:
Should we also mention the "config" object will be added if there is a task specific config in the `workflow_config`?